sky/execution.py

```diff
@@ -451,15 +456,47 @@ def launch(
         controller_utils.check_cluster_name_not_controller(
             cluster_name, operation_str='sky.launch')
 
+    handle = None
+    stages = None
+    # Check if cluster exists and we are doing fast provisioning
+    if fast and cluster_name is not None:
+        maybe_handle = global_user_state.get_handle_from_cluster_name(
+            cluster_name)
+        if maybe_handle is not None:
+            try:
+                # This will throw if the cluster is not available
+                backend_utils.check_cluster_available(
+                    cluster_name,
+                    operation='executing tasks',
+                    check_cloud_vm_ray_backend=False,
+                    dryrun=dryrun)
+                # If the cluster is available, restrict stages
+                handle = maybe_handle
+                stages = [
+                    # Stage.CLONE_DISK,
+                    # Stage.PROVISION,
+                    # Stage.OPTIMIZE,
+                    Stage.SYNC_WORKDIR,
+                    # Stage.SYNC_FILE_MOUNTS,
+                    # Stage.SETUP,
+                    # Stage.PRE_EXEC,
+                    Stage.EXEC,
+                    # Stage.DOWN
+                ]
+            except ClusterNotUpError:
```

Comment:
nit: [style guide](https://google.github.io/styleguide/pyguide.html#22-imports): import modules instead of directly importing classes 

`from sky import exceptions` 

```suggestion
            except exceptions.ClusterNotUpError:
```